artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQConsumer.java

```diff
@@ -380,6 +383,11 @@ public void setPrefetchSize(int prefetchSize) {
       this.prefetchSize = prefetchSize;
       this.currentWindow.set(prefetchSize);
       this.info.setPrefetchSize(prefetchSize);
+      if (this.prefetchSize == 0) {
+         messagePullHandler.compareAndSet(null, new MessagePullHandler());
```

Comment:
what happen if it won't succeed?

Reply:
@michaelandrepearce any suggestions?

Reply:
what do you mean if it wont succeed? if its not null it means it already had a pullhandler set, no need to assign a new one.

Reply:
that can happen if: 

1. there is a racing thread doing it 
2. the value isn't null

in both cases probably logging some trace/debug is ok if we expect it to happen?

Reply:
Anyway It's fine to me; if there are no objections I'm going to merge this ;)

Reply:
if its not null, we dont want to change it. e.g. the case there is client was already at prefetch 0 and simply sent a second update which set it to the same size, as such you want to leave the original pullhandler.

Reply:
if you really want a log i would add it as a trace, i dont think its needed tbh i think it just clutter the code base.

Reply:
no need to set messagePullHandler value if it isn't null

Reply:
:+1:  thanks for the explanation bud :)

Reply:
> Anyway It's fine to me; if there are no objections I'm going to merge this ;)

no objections here, please go ahead....